clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -349,19 +347,20 @@ private void sendProduceRequest(long now, int destination, short acks, int timeo
             recordsByPartition.put(tp, batch);
         }
 
-        ProduceRequest produceRequest = new ProduceRequest(acks, timeout, produceRecordsByPartition);
-        RequestHeader header = this.client.nextRequestHeader(ApiKeys.PRODUCE);
+        ProduceRequest.Builder requestBuilder =
+                new ProduceRequest.Builder(acks, timeout, produceRecordsByPartition);
```

Comment:
I think we need to enforce a minimum version of 2 since older versions would expect an older version of the message format. This is also an easy way to detect that we're talking with a version of the broker older than 0.10. I'm wondering if it would make sense to raise an exception to the user immediately when we connect to such a broker instead of waiting until we send an incompatible request?